File: src/tools/implementations/no_response.ts
import { RALRegistry } from "@/services/ral";
import type { AISdkTool, ToolExecutionContext } from "@/tools/types";
import { logger } from "@/utils/logger";
import { tool } from "ai";
import { z } from "zod";

const noResponseSchema = z.object({});

type NoResponseInput = z.infer<typeof noResponseSchema>;

interface NoResponseOutput {
    success: boolean;
    mode: "silent-complete";
    message: string;
}

async function executeNoResponse(
    _input: NoResponseInput,
    context: ToolExecutionContext
): Promise<NoResponseOutput> {
    const requested = RALRegistry.getInstance().requestSilentCompletion(
        context.agent.pubkey,
        context.conversationId,
        context.ralNumber
    );

    if (!requested) {
        throw new Error("Unable to request silent completion because the active RAL could not be found.");
    }

    logger.info("[no_response] Silent completion requested", {
        agent: context.agent.slug,
        conversationId: context.conversationId,
        ralNumber: context.ralNumber,
    });

    return {
        success: true,
        mode: "silent-complete",
        message:
            "Silent completion requested. This turn ends immediately with no assistant text, acknowledgement, emoji, or filler.",
    };
}

export function createNoResponseTool(context: ToolExecutionContext): AISdkTool {
    const aiTool = tool({
        description:
            "Request a silent completion for this turn. Use ONLY when the latest user message explicitly asks for no reply, including note-to-self or counting-aloud cases where the user does not want acknowledgements or filler. Calling this tool immediately ends the turn with no assistant text.",
        inputSchema: noResponseSchema,
        execute: async (input: NoResponseInput) => {
            return await executeNoResponse(input, context);
        },
    });

    Object.defineProperty(aiTool, "hasSideEffects", {
        value: true,
        enumerable: false,
        configurable: true,
    });

    return aiTool as AISdkTool;
}

File: crates/tenex-agent/src/tools/no_response.rs
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc,
};

use rig_core::{completion::ToolDefinition, tool::Tool};
use serde::{Deserialize, Serialize};
use serde_json::json;

#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct NoResponseError(String);

#[derive(Debug, Deserialize, Serialize)]
pub struct NoResponseArgs {}

#[derive(Debug, Serialize)]
pub struct NoResponseOutput {
    pub success: bool,
    pub mode: &'static str,
}

pub struct NoResponseTool {
    suppress: Arc<AtomicBool>,
}

impl NoResponseTool {
    pub fn new(suppress: Arc<AtomicBool>) -> Self {
        Self { suppress }
    }
}

impl Tool for NoResponseTool {
    const NAME: &'static str = "no_response";
    type Error = NoResponseError;
    type Args = NoResponseArgs;
    type Output = NoResponseOutput;

    async fn definition(&self, _prompt: String) -> ToolDefinition {
        ToolDefinition {
            name: Self::NAME.to_string(),
            description: "End this turn silently. Use ONLY when the latest user message explicitly \
                asks for no reply, including note-to-self or counting-aloud cases where the user \
                does not want acknowledgements or filler. Do not emit any text alongside this \
                call. Calling this tool ends the turn immediately — the agent loop stops and you \
                are not invoked again."
                .to_string(),
            parameters: json!({ "type": "object", "properties": {}, "required": [] }),
        }
    }

    async fn call(&self, _args: Self::Args) -> Result<Self::Output, Self::Error> {
        self.suppress.store(true, Ordering::Release);
        Ok(NoResponseOutput {
            success: true,
            mode: "silent-complete",
        })
    }
}
